js/script.js

```diff
@@ -1,120 +1,164 @@
-
-  if (searchTitle) {
-    const filteredTasks = tasks.filter((task) =>
-      task.title.toLowerCase().includes(searchTitle.toLowerCase())
-    );
-    if (filteredTasks.length > 0) {
-      renderTasks(filteredTasks);
+  const $overlay = showSpinnerOverlay($taskListContainer);
+
+  setTimeout(() => {
+    if (searchTitle) {
+      const filteredTasks = tasks.filter((task) =>
+        task.title.toLowerCase().includes(searchTitle.toLowerCase())
+      );
+      if (filteredTasks.length > 0) {
+        renderTasks(filteredTasks);
+      } else {
+        showToastMessage(MESSAGES.NO_TASKS_FOUND);
+      }
     } else {
-      showToastMessage("No tasks found matching the search.");
+      renderTasks(tasks);
     }
-  } else {
-    renderTasks(tasks);
-  }
-  $searchInput.value = "";
+
+    hideSpinnerOverlay($overlay);
+    $searchInput.value = "";
+  }, 1000);
 };
 
-const deleteTask = (taskId) => {
-  tasks = tasks.filter((task) => task.id !== taskId);
-  renderTasks(tasks);
+const deleteTask = (taskId, container) => {
+  const $overlay = showSpinnerOverlay(container);
+  setTimeout(() => {
+    tasks = tasks.filter((task) => task.id !== taskId);
+    renderTasks(tasks);
+    hideSpinnerOverlay($overlay);
+  }, 1000);
 };
 
 const editTask = (task) => {
-  task.isEditing = true;
+  task.editMode = true;
   renderTasks(tasks);
 };
 
-const updateTask = (task, newTitle) => {
+const updateTask = (task, container, newTitle) => {
   if (newTitle) {
-    task.title = newTitle;
-    task.isEditing = false;
+    const $overlay = showSpinnerOverlay(container);
+    setTimeout(() => {
+      task.title = newTitle;
+      task.editMode = false;
+      renderTasks(tasks);
+      hideSpinnerOverlay($overlay);
+    }, 1000);
   }
-  renderTasks(tasks);
 };
 
-const completeTask = (taskId) => {
-  const task = tasks.find((task) => task.id === taskId);
-
-  if (task === undefined) {
-    return;
-  }
-
-  task.done = true;
-  task.isEditing = false;
+const completeTask = (taskId, container) => {
+  const $overlay = showSpinnerOverlay(container);
+  setTimeout(() => {
+    const task = tasks.find((task) => task.id === taskId);
+    if (task === undefined) {
+      return;
+    }
 
-  renderTasks(tasks);
+    task.done = true;
+    task.isEditing = false;
+    renderTasks(tasks);
+    hideSpinnerOverlay($overlay);
+  }, 1000);
 };
 
 const createTask = (taskTitle) => {
   const task = {
     id: new Date().getTime(),
     title: taskTitle,
     createdAt: formatDate(new Date()),
+    editMode: false,
+    done: false,
   };
   tasks.unshift(task);
   renderTasks(tasks);
-  $taskInput.value = "";
 };
 
-const renderTasks = (tasks = []) => {
-  $taskList.innerHTML = "";
-
-  tasks.forEach((task) => {
-    const $taskElement = createTaskElement(
-      task,
-      deleteTask,
-      editTask,
-      updateTask,
-      cancelEdit,
-      completeTask
-    );
-    $taskList.appendChild($taskElement);
+const renderTasks = (tasksToRender = tasks) => {
```

Comment:
why tasks is not an empty array now?

Reply:
Here I have planned to pass filtered tasks or tasks array. If filtered array passed it will render the filtered tasks or the whole tasks array. That's why didn't keep default empty array.